contracts/base/strategies/alpaca/AlpacaVaultBase.sol

```diff
@@ -0,0 +1,171 @@
+  /// @dev Should be incremented when contract changed
+  string public constant VERSION = "1.0.0";
+  /// @dev Placeholder, for non full buyback need to implement liquidation
+  uint private constant _BUY_BACK_RATIO = 1000;
+  IAlpacaVault private alpacaVault;
+  IFairLaunch private fairLaunchPool;
+  uint private poolID;
+
+  /// @notice Contract constructor using on strategy implementation
+  /// @dev The implementation should check each parameter
+  /// @param _controller Controller address
+  /// @param _vault SmartVault address that will provide liquidity
+  /// @param _underlying Underlying token address
+  /// @param __rewardTokens Reward tokens that the strategy will farm
+  /// @param _alpacaVault AlpacaVault address
+  /// @param _fairLaunch FairLaunch address
+  /// @param _poolId alpaca Pool id
+  constructor(
+    address _controller,
+    address _vault,
+    address _underlying,
+    address[] memory __rewardTokens,
+    address _alpacaVault,
+    address _fairLaunch,
+    uint _poolId
+  ) StrategyBase(_controller, _underlying, _vault, __rewardTokens, _BUY_BACK_RATIO) {
+    alpacaVault = IAlpacaVault(_alpacaVault);
+    fairLaunchPool = IFairLaunch(_fairLaunch);
+    poolID = _poolId;
+    require(alpacaVault.token() == _underlyingToken, "CSB: Wrong underlying");
+  }
+
+  // ************* VIEWS *******************
+
+  /// @notice Strategy balance in the fairLaunch pool
+  /// @return bal Balance amount in underlying tokens
+  function rewardPoolBalance() public override view returns (uint) {
+    uint totalSupply = alpacaVault.totalSupply();
+    uint totalToken = alpacaVault.totalToken();
+    (uint amount,) = fairLaunchPool.userInfo(poolID, address(this));
+    uint balance = amount * totalToken / totalSupply;
+    return balance;
+  }
+
+  /// @notice Return approximately amount of reward tokens ready to claim in fairLaunch pool
+  /// @dev Don't use it in any internal logic, only for statistical purposes
+  /// @return Array with amounts ready to claim
+  function readyToClaim() external view override returns (uint[] memory) {
+    uint[] memory toClaim = new uint[](1);
+    toClaim[0] = fairLaunchPool.pendingAlpaca(poolID, address(this));
+    return toClaim;
+  }
+
+  /// @notice TVL of the underlying in the fairLaunch pool
+  /// @dev Only for statistic
+  /// @return Pool TVL
+  function poolTotalAmount() external view override returns (uint) {
+    uint totalSupply = alpacaVault.totalSupply();
+    uint totalToken = alpacaVault.totalToken();
+    uint fairLaunchPoolBalance = IERC20(address(alpacaVault)).balanceOf(address(fairLaunchPool));
+    return fairLaunchPoolBalance  * totalToken / totalSupply;
+  }
+
+  // ************ GOVERNANCE ACTIONS **************************
+
+  /// @notice Claim rewards from external project and send them to FeeRewardForwarder
+  function doHardWork() external onlyNotPausedInvesting override hardWorkers {
+    depositToPool(IERC20(_underlyingToken).balanceOf(address(this)));
```

Comment:
use investAllUnderlying()
it will check zero balance

Reply:
Done